A74-xds-config-tears.md

```diff
@@ -0,0 +1,311 @@
+```
+
+Note that the watcher's OnError() and OnResourceDoesNotExist()
+methods will be invoked only for problems with the listener and route
+configuration resources.  For cluster-specific errors, we need to
+retain the existing behavior, which is that it causes a problem for the
+individual cluster but does not affect traffic sent to any other cluster.
+Therefore, when the XdsDependencyManager sees an error on a cluster for
+which it did not previously have data, or if it sees a does-not-exist
+notification on a cluster, it will report that by returning a non-OK
+status for the cluster in the XdsConfig.  Similarly, if there is a problem
+obtaining endpoint data from EDS or DNS, the XdsDependencyManager will
+report that problem via the cluster's resolution_note field.
+
+The xds resolver will pass the returned XdsConfig to the LB policies via
+an attribute, so that the LB policies can use the cluster data.  Note
+that we cannot simply pass the data via the cds LB policy's JSON config,
+because we need this to work in cases where the cluster is determined
+dynamically, such as where a route uses a ClusterSpecifierPlugin.  (We
+have some internal functionality that uses this, not something that has
+been documented in a public gRFC.)
+
+In order to handle such cases, the XdsDependencyManager will also provide
+an API to subscribe to a particular cluster resource whose name was
+not present in the route configuration.  The xds resolver will pass the
+XdsDependencyManager to the LB policies via an attribute, so that the
+cds LB policy can use this API.  The API will return some subscription
+handle that the caller can release when they are no longer interested
+in the cluster subscription.
+
+Note that in the XdsConfig, the cluster map will contain an entry for
+every cluster that is either referenced in the route configuration or
+for which a dynamic subscription was registered.  If a cluster is
+present only underneath an aggregate cluster, it will be present in the
+list for that aggregate cluster, but it will not have its own top-level
+entry.
+
+#### Changes in cds LB policy
+
+Instead of starting a watch on the XdsClient to get the cluster data,
+the cds LB policy will get its configuration by lookup up the cluster
+name in the XdsConfig passed to it via an attribute.
+
+To address the case where the cluster is determined dynamically (e.g., a
+route that uses a ClusterSpecifierPlugin), we will add a new field to
+the cds LB policy's configuration:
+
+```
+// Set this to true for a cluster that is determined dynamically
+// rather than being present in the route configuration.
+bool is_dynamic = 2;
+```
+
+If that field is set to true, then if the cluster name is not present in
+the XdsConfig, the cds policy will obtain a subscription handle from the
+XdsDependencyManager if it has not already done so.  This handle will
+not be released until the cds policy is destroyed.
+
+Note that when XdsDependencyManager returns a new configuration that
```

Comment:
I had the same idea when I was implementing this, but when I looked at the details, there were enough differences that it didn't really make sense to use a single mechanism for both.

The purpose of the cluster refs taken by the ConfigSelector is to make sure that we don't remove the cluster from the LB policy tree generated by the xds resolver if the cluster gets removed from the RouteConfig before the calls have finished their LB picks.  It's true that prior to this design, retaining the cds policy also had the side effect of retaining the CDS watch, but I'm not sure we actually need that -- in general, these refs should just keep the cds policy around for a fairly short amount of time, so it probably doesn't matter if we stop the CDS watch and just keep the policy around with its last-seen config, because there probably won't be any updates in that short window that we couldn't just ignore.  (Having just written this out, though, I would be open to an argument about edge cases that may make this window longer than we would normally expect, where we might not want to ignore subsequent CDS updates.)

In contrast, the clusters that are dynamically created by the ClusterSpecifierPlugin are never going to be present in the LB policy config generated by the xds resolver in the first place.  The purpose of the dynamic subscription is specifically to trigger the existence of the CDS watch, because unlike the ConfigSelector case, here we do need the CDS watch to exist for as long as the ClusterSpecifierPlugin decides to keep the cds child policy alive.

Reply:
As per our conversation today, I've changed this such that the cluster refs held by the ConfigSelector will also hold a subscription from the XdsDependencyManager, so we'll ensure that the CDS watch sticks around as long as the cluster is in the xds_cluster_manager LB policy config.